tests/unit/s2n_fork_generation_number_test.c

```diff
@@ -0,0 +1,349 @@
+
+    /* Some fork detection mechanisms can also detect forks through clone() */
+    if (test_case->test_case_must_pass_clone_test == CLONE_TEST_YES) {
+        EXPECT_EQUAL(s2n_assert_madv_wipeonfork_is_supported() || s2n_assert_map_inherit_zero_is_supported(), true);
+        EXPECT_EQUAL(s2n_unit_test_clone(return_fork_generation_number), S2N_SUCCESS);
+    }
+    else if (test_case->test_case_must_pass_clone_test == CLONE_TEST_DETERMINE_AT_RUNTIME) {
+        if (s2n_assert_madv_wipeonfork_is_supported() == true ||
+            s2n_assert_map_inherit_zero_is_supported() == true) {
+            EXPECT_EQUAL(s2n_unit_test_clone(return_fork_generation_number), S2N_SUCCESS);
+        }
+    }
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_test_case_default_cb(struct fgn_test_case *test_case)
+{
+    EXPECT_EQUAL(s2n_unit_tests_common(test_case), S2N_SUCCESS);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_test_case_pthread_atfork_cb(struct fgn_test_case *test_case)
+{
+    POSIX_GUARD_RESULT(s2n_ignore_wipeonfork_and_inherit_zero_for_testing());
+    EXPECT_EQUAL(s2n_unit_tests_common(test_case), S2N_SUCCESS);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_test_case_madv_wipeonfork_cb(struct fgn_test_case *test_case)
+{
+    if (s2n_assert_madv_wipeonfork_is_supported() == false) {
+        FGN_TEST_CASE_PRINT_MSG_INFO("Test case not supported. Skipping.", test_case);
+        return S2N_SUCCESS;
+    }
+
+    POSIX_GUARD_RESULT(s2n_ignore_pthread_atfork_for_testing());
+    EXPECT_EQUAL(s2n_unit_tests_common(test_case), S2N_SUCCESS);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_test_case_map_inherit_zero_cb(struct fgn_test_case *test_case)
+{
+    if (s2n_assert_map_inherit_zero_is_supported() == false) {
+        FGN_TEST_CASE_PRINT_MSG_INFO("Test case not supported. Skipping.", test_case);
+        return S2N_SUCCESS;
+    }
+
+    POSIX_GUARD_RESULT(s2n_ignore_pthread_atfork_for_testing());
+    EXPECT_EQUAL(s2n_unit_tests_common(test_case), S2N_SUCCESS);
+
+    return S2N_SUCCESS;
+}
+
+struct fgn_test_case fgn_test_cases[NUMBER_OF_FGN_TEST_CASES] = {
+    {"Default fork detect mechanisms.", s2n_test_case_default_cb, CLONE_TEST_DETERMINE_AT_RUNTIME},
+    {"Only pthread_atfork fork detection mechanism.", s2n_test_case_pthread_atfork_cb, CLONE_TEST_NO},
+    {"Only madv_wipeonfork fork detection mechanism.", s2n_test_case_madv_wipeonfork_cb, CLONE_TEST_YES},
+    {"Only map_inheret_zero fork detection mechanism.", s2n_test_case_map_inherit_zero_cb, CLONE_TEST_YES}
+};
+/* s2n_array_len is run-time but standard static array length calculation is
+ * captured by the simple errors script. However, the script is perfoming static
+ * analysis on the pure source code, so, we can easily trick it :).
+ */
+#define SIZE_OF_COPY(x) sizeof(x)
+S2N_STATIC_ASSERT(NUMBER_OF_FGN_TEST_CASES == (SIZE_OF_COPY(fgn_test_cases) / SIZE_OF_COPY(fgn_test_cases[0])));
```

Comment:
Why does this need to be a static assert? Couldn't you just do a standard assertion in main()?

Reply:
Hi, performance!
But, yes, we could certainly do that.